### ISSUE TITLE
App defined share permission flags

### ISSUE THREAD

**Issue by juliushaertl** [nextcloud/server#22220]
Moved over from https://github.com/nextcloud/server/issues/12178#issuecomment-528338333 since this recently came up again with ONLYOFFICE.

There are several use-cases where apps like Collabora or OnlyOffice could provide additional sharing permissions. 

OnlyOffice (https://api.onlyoffice.com/editors/config/document/permissions):
- Can fill forms
- Can review
- Can comment
- Can print (would be covered by hide download)
- Can download (would be covered by hide download)
- Can edit (would be covered by regular permissions)

Collabora has quite fine granular options to set (DisableCopy, DisableExport, HidePrintOption, HideSaveOption, HideExportOption, UserCanWrite) but I think we do fine by just using our current hide download/read/write permissions for those. However in the future there also might be options similar to the ones in OnlyOffice.

Edit: It should be possible for apps to limit the availability of permissions like https://api.onlyoffice.com/editors/faq/sharing

Additional options that might be possible for link shares:
- Watermark file (only makes sense for hide download links, as otherwise you can always just download the regular file), so I think it's fine to not expose this as an additional permission

My proposal would be to allow apps adding custom permission flags that can either be true or false and then would be added to the additional settings

I've also did some quick mockups to make the menu restructuring a bit clearer:
![image](https://user-images.githubusercontent.com/3404133/64341420-98892d80-cfe8-11e9-8d30-92269ac48b11.png)

Additional permissions would then pile up under the Can delete, only if the shared mime type would support them of course.

**Comment by juliushaertl** [comment]
@nextcloud/sharing @nextcloud/designers

**Comment by rullzer** [comment]
So (API wise aside etc). You are suggesting to have an extra field in the share table. And store some JSON thing there with a couple of bools?

Should be fine in my opinion.

**Comment by juliushaertl** [comment]
> So (API wise aside etc). You are suggesting to have an extra field in the share table. And store some JSON thing there with a couple of bools?

Either that or we have a extended share permission table that we can then query if needed only

**Comment by szaimen** [comment]
I suppose this is still valid?

**Comment by skjnldsv** [comment]
@jancborchardt fixed witht he concept of the new sharing ui, right?

**Comment by juliushaertl** [comment]
No that would be a new feature that gets requested from time to time to give a more granular permission options is specific collaborative editing solutions like ONLYOFFICE or Collabora Online.

**Comment by susnux** [comment]
something for unified sharing? cc @provokateurin

**Comment by provokateurin** [comment]
Yes, I think we can close this issue in favor of https://github.com/nextcloud/server/issues/51803, as the current approach will allow apps to define custom permissions :)

**Comment by jancborchardt** [comment]
Closing then for cleanup :)
